src/core/SampleBufferCache.cpp

```diff
@@ -0,0 +1,84 @@
+/*
+ * SampleBufferCache.cpp - Cache for storing SampleBuffer's
+ *
+ * Copyright (c) 2024 saker <sakertooth@gmail.com>
+ *
+ * This file is part of LMMS - https://lmms.io
+ *
+ * This program is free software; you can redistribute it and/or
+ * modify it under the terms of the GNU General Public
+ * License as published by the Free Software Foundation; either
+ * version 2 of the License, or (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
+ * General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public
+ * License along with this program (see COPYING); if not, write to the
+ * Free Software Foundation, Inc., 51 Franklin Street, Fifth Floor,
+ * Boston, MA 02110-1301 USA.
+ *
+ */
+
+#include "SampleBufferCache.h"
+
+#include <cassert>
+#include <cerrno>
+#include <stdexcept>
+
+namespace lmms {
+
+SampleBufferCache::SampleBufferCache(size_t capacity)
+	: m_capacity(capacity)
+{
+}
+
+void SampleBufferCache::add(const QString& key, Buffer buffer)
+{
+	if (buffer == nullptr) { throw std::runtime_error{"Cannot add null buffer"}; }
+
+	if (m_cache.contains(key))
+	{
+		throw std::runtime_error{std::string{"Buffer with the same key already exists within the cache"}};
+	}
+
+	assert(m_cache.size() <= m_capacity);
+	if (m_cache.size() == m_capacity)
+	{
+		const auto leastRecentlyUsed = std::min_element(m_cache.begin(), m_cache.end(),
+			[](const auto& entryOne, const auto& entryTwo) { return entryOne.use_count() < entryTwo.use_count(); });
+		if (leastRecentlyUsed == m_cache.end()) { throw std::runtime_error{"Failed to evict buffer from cache"}; }
```

Comment:
I think the only case where this can happen is `m_cache` is empty(`m_capacity == 0` or some race conditions). A cache with zero capacity doesn't make sense, so could we enforce non-zero capacity instead?